cpp/src/arrow/array/builder_encoded.cc

```diff
@@ -0,0 +1,213 @@
+  }
+  return value_builder().AppendArraySlice(ree_util::ValuesArray(to_append),
+                                          physical_offset, physical_length);
+}
+
+Status RunEndEncodedBuilder::AppendArraySlice(const ArraySpan& array, int64_t offset,
+                                              int64_t length) {
+  // Finish eventual runs started using AppendScalars() and others before. We don't
+  // attempt to merge them with the runs from the appended array slice for now
+  RETURN_NOT_OK(FinishRun());
+
+  ARROW_DCHECK(offset + length <= array.length);
+  ARROW_DCHECK(array.type->Equals(type_));
+  // Create a slice of the slice for the part we actually want to add
+  ArraySpan to_append = array;
+  to_append.SetSlice(array.offset + offset, length);
+
+  switch (type_->run_ends_type()->id()) {
+    case Type::INT16:
+      RETURN_NOT_OK(DoAppendArray<int16_t>(to_append));
+      break;
+    case Type::INT32:
+      RETURN_NOT_OK(DoAppendArray<int32_t>(to_append));
+      break;
+    case Type::INT64:
+      RETURN_NOT_OK(DoAppendArray<int64_t>(to_append));
+      break;
+    default:
+      return Status::Invalid("Invalid type for run ends array: ", type_->run_ends_type());
+  }
+
+  // next run is not merged either
+  run_start_ = length_;
+  return Status::OK();
+}
+
+std::shared_ptr<DataType> RunEndEncodedBuilder::type() const { return type_; }
+
+Status RunEndEncodedBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
+  ARROW_ASSIGN_OR_RAISE(auto run_ends_array, run_end_builder().Finish());
+  ARROW_ASSIGN_OR_RAISE(auto values_array, value_builder().Finish());
+  ARROW_ASSIGN_OR_RAISE(auto ree_array,
+                        RunEndEncodedArray::Make(run_ends_array, values_array, length_));
+  *out = std::move(ree_array->data());
+  return Status::OK();
+}
+
+template <typename RunEndsType>
+Status RunEndEncodedBuilder::DoAppendRunEnd() {
+  return internal::checked_cast<typename CTypeTraits<RunEndsType>::BuilderType*>(
+             children_[0].get())
+      ->Append(static_cast<RunEndsType>(length_));
+}
+
+Status RunEndEncodedBuilder::FinishRun() {
+  if (length_ - run_start_ == 0) {
+    return Status::OK();
+  }
+  if (current_value_) {
+    RETURN_NOT_OK(value_builder().AppendScalar(*current_value_));
+  } else {
+    RETURN_NOT_OK(value_builder().AppendNull());
+  }
+  switch (type_->run_ends_type()->id()) {
+    case Type::INT16:
+      RETURN_NOT_OK(DoAppendRunEnd<int16_t>());
+      break;
+    case Type::INT32:
+      RETURN_NOT_OK(DoAppendRunEnd<int32_t>());
+      break;
+    case Type::INT64:
+      RETURN_NOT_OK(DoAppendRunEnd<int64_t>());
+      break;
+    default:
+      return Status::Invalid("Invalid type for run ends array: ", type_->run_ends_type());
+  }
+  current_value_.reset();
+  run_start_ = 0;
+  return Status::OK();
+}
+
+Status RunEndEncodedBuilder::ResizePhyiscal(int64_t capacity) {
+  RETURN_NOT_OK(value_builder().Resize(capacity));
+  return run_end_builder().Resize(capacity);
```

Comment:
Why resize both builders with the same capacity? This doesn't seem to make sense.

Reply:
How else would it be? For every `i < physical_length` `run_ends[i]` is the end of the run of `values[i]`.